### PR TITLE
deal.II: Further modernisation and improvements

### PR DESCRIPTION
**Pull request by jppelteret**

### CODE REVIEW

**Review by adamjstewart — APPROVED** [review]

**Comment by jppelteret** [comment]
This is a relatively straight forward set of updates, so hopefully it won't take much effort to review. I was able to find more places to use `self.define_from_variant`, which is great. I spent the afternoon building the package from scratch with as many dependencies as possible. Between this and #19137 everything seems to be working fine.

File: var/spack/repos/builtin/packages/dealii/package.py [hunk 1]
```diff
@@ -483,30 +470,26 @@ def cmake_args(self):
...
                 '-DSCALAPACK_FOUND=true',
                 '-DSCALAPACK_INCLUDE_DIRS=%s' % (
                     spec['scalapack'].prefix.include),
-                '-DSCALAPACK_LIBRARIES=%s' % scalapack.joined(';'),
-                '-DDEAL_II_WITH_SCALAPACK=ON'
-            ])
-        else:
-            options.extend([
-                '-DDEAL_II_WITH_SCALAPACK=OFF'
+                '-DSCALAPACK_LIBRARIES=%s' % scalapack.joined(';')
```
Comment (sethrj) [inline var/spack/repos/builtin/packages/dealii/package.py]: The static `CMakePackage.define(key, value)` method formats the `value` according to the type: so if you do
```suggestion
                self.define('SCALAPACK_LIBRARIES', scalapack)
```
it'll do the list-to-cmake conversions for you. Likewise with boolean arguments. There are a couple of other places you could to this if you want.

Comment (jppelteret) [reply var/spack/repos/builtin/packages/dealii/package.py]: Thanks for the suggestion! It's great that the CMake syntax is now taken care of by the package. There were a lot of changes that came from employing this approach, so I'm going to need a little time to check that there are no negative consequences from using it.

File: var/spack/repos/builtin/packages/dealii/package.py [hunk 2]
```diff
@@ -274,61 +274,68 @@ class Dealii(CMakePackage, CudaPackage):
...
+            self.define('DEAL_II_ALLOW_BUNDLED', False)
         ])
 
         if spec.satisfies('@:8.99'):
             options.extend([
                 # Cmake may still pick up system's bzip2, fix this:
-                '-DBZIP2_FOUND=true',
-                '-DBZIP2_INCLUDE_DIRS=%s' % spec['bzip2'].prefix.include,
-                '-DBZIP2_LIBRARIES=%s' % spec['bzip2'].libs.joined(';')
+                self.define('BZIP2_FOUND', True.directories),
```
Comment (sethrj) [inline var/spack/repos/builtin/packages/dealii/package.py]: Copy-paste error

Comment (jppelteret) [reply var/spack/repos/builtin/packages/dealii/package.py]: Good catch! Thanks for noticing this.

**Comment by jppelteret** [comment]
@luca-heltai FYI. Another set of modernisation changes. As best I can tell, the configure arguments that are passed to CMake are still correct. The only noticeable difference should be that the type (bool, string, etc.) is now qualified for each option.
```
-DDEAL_II_EXAMPLES_RELDIR:STRING=share/deal.II/examples 
-DDEAL_II_DOCREADME_RELDIR:STRING=share/deal.II/ 
-DDEAL_II_DOCHTML_RELDIR:STRING=share/deal.II/doc 
-DBOOST_DIR:STRING=<PATH>/boost-1.74.0-pa6dyhiwggv5jjvnqp53skpv7l6f5xqv 
-DLAPACK_FOUND:BOOL=ON 
-DLAPACK_INCLUDE_DIRS:STRING=<PATH>/openblas-0.3.10-vb6zb2n4ljbwmcxzlxz72i3owca5zhb3/include 
-DLAPACK_LIBRARIES:STRING=<PATH>/openblas-0.3.10-vb6zb2n4ljbwmcxzlxz72i3owca5zhb3/lib/libopenblas.dylib 
-DUMFPACK_DIR:STRING=<PATH>/suite-sparse-5.7.2-xrv6hsnd4wsrlk34qtzjo7ci5jehlbwg 
-DZLIB_DIR:STRING=<PATH>/zlib-1.2.11-vb3iyrgf7q2w5ifl22oxegblzjxgee45 
-DDEAL_II_ALLOW_BUNDLED:BOOL=OFF 
-DDEAL_II_COMPONENT_DOCUMENTATION:BOOL=OFF 
-DDEAL_II_COMPONENT_EXAMPLES:BOOL=OFF 
-DDEAL_II_WITH_64BIT_INDICES:BOOL=OFF 
-DDEAL_II_WITH_CUDA:BOOL=OFF 
-DDEAL_II_WITH_MPI:BOOL=ON 
-DCMAKE_C_COMPILER:STRING=<PATH>/openmpi-3.1.6-u76x7zshzem2krvt75curyvqs2qng3nk/bin/mpicc 
-DCMAKE_CXX_COMPILER:STRING=<PATH>/openmpi-3.1.6-u76x7zshzem2krvt75curyvqs2qng3nk/bin/mpic++ 
-DCMAKE_Fortran_COMPILER:STRING=<PATH>/openmpi-3.1.6-u76x7zshzem2krvt75curyvqs2qng3nk/bin/mpif90 
-DMPI_C_COMPILER:STRING=<PATH>/openmpi-3.1.6-u76x7zshzem2krvt75curyvqs2qng3nk/bin/mpicc 
-DMPI_CXX_COMPILER:STRING=<PATH>/openmpi-3.1.6-u76x7zshzem2krvt75curyvqs2qng3nk/bin/mpic++ 
-DMPI_Fortran_COMPILER:STRING=<PATH>/openmpi-3.1.6-u76x7zshzem2krvt75curyvqs2qng3nk/bin/mpif90 
-DDEAL_II_COMPONENT_PYTHON_BINDINGS:BOOL=OFF 
-DDEAL_II_WITH_THREADS:BOOL=ON 
-DTBB_DIR:STRING=<PATH>/intel-tbb-2020.3-tkcnwb6cc3h7htfdne4zzow2jtmjun5a 
-DDEAL_II_WITH_GSL:BOOL=ON 
-DGSL_DIR:STRING=<PATH>/gsl-2.5-qq47d4lneh34d7p2sq7vblwgz2mnbbgp
-DDEAL_II_WITH_HDF5:BOOL=ON 
-DHDF5_DIR:STRING=<PATH>/hdf5-1.10.7-jpsfkiotpg3ye4dkbwlalvbv2gqtgeby 
-DDEAL_II_WITH_P4EST:BOOL=ON 
-DP4EST_DIR:STRING=<PATH>/p4est-2.2-bxlej7nvjdye7dplqt2htcijr2db7fkl 
-DDEAL_II_WITH_PETSC:BOOL=ON
-DPETSC_DIR:STRING=<PATH>/petsc-3.14.0-5n7xosftkf7jkcpvr4lv7harsj62e7gj 
-DDEAL_II_WITH_SLEPC:BOOL=ON 
-DSLEPC_DIR:STRING=<PATH>/slepc-3.14.0-lnw5fr3zznzfiyp7helrqh3onall5kdk 
-DDEAL_II_WITH_TRILINOS:BOOL=ON 
-DTRILINOS_DIR:STRING=<PATH>/trilinos-12.10.1-atbbkm2tyjmlqvxljxw5fqo6zb275rqo 
-DDEAL_II_WITH_METIS:BOOL=ON 
-DMETIS_DIR:STRING=<PATH>/metis-5.1.0-u4etlkmanldltanys6aw6yfx74rknjnp 
-DDEAL_II_WITH_SUNDIALS:BOOL=ON 
-DSUNDIALS_DIR:STRING=<PATH>/sundials-3.2.1-23aadrd6uu6bjda7hzsr2sl5hcpbyji7 
-DDEAL_II_WITH_NANOFLANN:BOOL=ON 
-DNANOFLANN_DIR:STRING=<PATH>/nanoflann-1.2.3-23zb5nkpvajhpgljjccsuerauiikcdhj 
-DDEAL_II_WITH_ASSIMP:BOOL=ON 
-DASSIMP_DIR:STRING=<PATH>/assimp-4.0.1-vv4eytsn7plzieizfu7gfawiqct25lbc 
-DDEAL_II_WITH_GMSH:BOOL=ON 
-DGMSH_DIR:STRING=<PATH>/gmsh-4.5.4-ix5qzsn5hmrp4l3znvvap4hv6xyixkfg 
-DDEAL_II_WITH_MUPARSER:BOOL=ON 
-DMUPARSER_DIR:STRING=<PATH>/muparser-2.2.6.1-ovdzcuiq2sgmmvnbeab76jxwqxssquby 
-DDEAL_II_WITH_SYMENGINE:BOOL=ON 
-DSYMENGINE_DIR:STRING=<PATH>/symengine-0.6.0-4lptpry57sfiy4cl3irsqy5upjyxx5q4 
-DDEAL_II_WITH_GINKGO:BOOL=OFF 
-DDEAL_II_WITH_ADOLC:BOOL=ON 
-DADOLC_DIR:STRING=<PATH>/adol-c-2.7.2-acalecuckmczenspldqtc4mzb3sh4vxl 
-DDEAL_II_WITH_ARPACK:BOOL=ON 
-DARPACK_DIR:STRING=<PATH>/arpack-ng-3.7.0-pldodwivqfkfueq7jlgjk2nrtfqnauxc 
-DDEAL_II_ARPACK_WITH_PARPACK:BOOL=ON 
-DDEAL_II_WITH_NETCDF:BOOL=OFF 
-DDEAL_II_WITH_SCALAPACK:BOOL=ON 
-DSCALAPACK_FOUND:BOOL=ON 
-DSCALAPACK_INCLUDE_DIRS:STRING=<PATH>/netlib-scalapack-2.1.0-m2etjmhf35e5rrtozwdxitl5fulp75kz/include 
-DSCALAPACK_LIBRARIES:STRING=<PATH>/netlib-scalapack-2.1.0-m2etjmhf35e5rrtozwdxitl5fulp75kz/lib/libscalapack.dylib 
-DDEAL_II_WITH_OPENCASCADE:BOOL=ON 
-DOPENCASCADE_DIR:STRING=<PATH>/oce-0.18.3-rgzqyffh6ztolkvzbyixpqykzfmpthjj 
-DDEAL_II_CXX_FLAGS:STRING=
```

File: var/spack/repos/builtin/packages/dealii/package.py [hunk 3]
```diff
@@ -343,19 +350,19 @@ def cmake_args(self):
...
-            'DEAL_II_WITH_64BIT_INDICES', 'int64')
-        )
+            'DEAL_II_WITH_64BIT_INDICES', 'int64'
+        ))
 
         if (spec.satisfies('^openblas+ilp64') or
             spec.satisfies('^intel-mkl+ilp64') or
             spec.satisfies('^intel-parallel-studio+mkl+ilp64')):
-            options.append('-DLAPACK_WITH_64BIT_BLAS_INDICES=ON')
+            self.define('LAPACK_WITH_64BIT_BLAS_INDICES', True)
```
Comment (sethrj) [inline var/spack/repos/builtin/packages/dealii/package.py]: `options.append` is missing here: the line is a no-op.

**Review by sethrj — COMMENTED** [review]
One more typo fix and this should be good to go!

File: var/spack/repos/builtin/packages/dealii/package.py (hunk 3 above)
Comment (jppelteret) [reply var/spack/repos/builtin/packages/dealii/package.py]: Thanks for spotting this error as well. I've fixed it up!

**Comment by jppelteret** [comment]
Great, thanks a lot @sethrj and @adamjstewart!

**Pull request merged**